pact_testgen/broker.py

```diff
@@ -0,0 +1,93 @@
+import urllib
+from typing import Any, Dict, Optional, Tuple
+
+import requests
+from pydantic import BaseSettings, Field
+from pydantic.env_settings import InitSettingsSource, SettingsSourceCallable
+
+from .models import Pact
+
+
+# For broker settings, we want to allow passing None for username
+# and password, and revert to loading from the environment. However,
+# we also want to prioritize username and password if they are passed
+# as init kwargs, but are not None. This pattern simplifies CLI option
+# handling, at the expense of additional complexity here.
+
+
+class InitSettingsSourceIgnoreNone(InitSettingsSource):
+    """
+    An InitSettingsSource that ignores arguments
+    with a value of None.
+    """
+
+    def __init__(self, init_kwargs: Dict[str, Any]):
+        self.init_kwargs = {k: v for k, v in init_kwargs.items() if v is not None}
+        print(f"InitSettingsSourceIgnoreNone.init_kwargs={self.init_kwargs}")
+
+
+class BrokerConfig(BaseSettings):
+    # Same env vars as pact broker CLI client
+    # https://github.com/pact-foundation/pact_broker-client#usage---cli
+    base_url: str
+    username: str = Field(default="")
+    password: str = Field(default="")
+
+    class Config:
+        env_prefix = "pact_broker_"
+
+        @classmethod
+        def customise_sources(
+            cls,
+            init_settings: SettingsSourceCallable,
+            env_settings: SettingsSourceCallable,
+            file_secret_settings: SettingsSourceCallable,
+        ) -> Tuple[SettingsSourceCallable, ...]:
+            return (
+                InitSettingsSourceIgnoreNone(init_settings.init_kwargs),
+                env_settings,
+                file_secret_settings,
+            )
+
+    @property
+    def auth_tuple(self) -> Optional[Tuple[str, str]]:
+        """
+        Returns auth value suitable for passing to requests.
+
+        If username or password are present, will be Tuple[str,str],
+        else None.
+        """
+        if self.username or self.password:
+            return (self.username, self.password)
```

Comment:
❓ Presumably the broker will need both if it requires authentication at all. Is it worth failing loudly if either aren't provided at this point?

Reply:
I think technically the basic auth spec allows empty username or password. That said, after some quick testing with the pact broker docker image, it seems the pact broker doesn't enable basic auth unless both are set, even if `pact_broker_enable_basic_auth` is explicitly set to `true`. So I think you're right that we can expect to see both here.

Reply:
Because the pact broker actually supports both basic + token auth<sup>1</sup>, and multi-param validation is a bit awkward in pydantic, I'm trying to set up a `BrokerBasicAuthConfig` nested settings class, something like

```py

class BrokerBasicAuthConfig(BaseSettings):
    username: str
    password: str


class BrokerConfig(BaseSettings):
    auth: Optional[BrokerBasicAuthConfig]
    ...
```

which could later be expanded to something like

```py
class BrokerConfig(BaseSettings):
    auth: Union[None, BrokerBasicAuthConfig, BrokerTokenAuthConfig]
    ...
```

but for whatever reason it's completely blowing up in my face with errors like `AttributeError: '_SpecialForm' object has no attribute 'username'`.

I did verify with a simple test case that pydantic does in fact support nested `BaseSettings` classes, so IDK what's going on. Might punt on it + create an issue for tracking.


<sup>1</sup>Token auth only supported in the paid-for Pactflow